pallets/circuit/src/lib.rs

```diff
@@ -1456,133 +1554,110 @@ impl<T: Config> Pallet<T> {
+                return Err("Side_effect max_reward must be equal to reward of Optional Insurance")
             }
+            if sfx.insurance != insurance {
+                return Err("Side_effect insurance must be equal to reward of Optional Insurance")
+            }
+            let submission_target_height = T::Portal::get_latest_finalized_height(sfx.target)?
+                .ok_or("target height not found")?;
+
+            full_side_effects.push(FullSideEffect {
+                input: sfx.clone(),
+                confirmed: None,
+                security_lvl: determine_security_lvl(gateway_type),
+                submission_target_height,
+                best_bid: None,
+            });
         }
-
         // Circuit's automatic side effect ordering: execute escrowed asap, then line up optimistic ones
         full_side_effects.sort_by(|a, b| b.security_lvl.partial_cmp(&a.security_lvl).unwrap());
 
         let mut full_side_effects_steps: Vec<
             Vec<FullSideEffect<T::AccountId, T::BlockNumber, EscrowedBalanceOf<T, T::Escrowed>>>,
-        > = vec![vec![]];
-
-        for sorted_fse in full_side_effects {
-            let current_step = full_side_effects_steps
-                .last_mut()
-                .expect("Vector initialized at declaration");
-
-            // Push to the single step as long as there's no Dirty side effect
-            //  Or if there was no Optimistic/Escrow side effects before
-
-            if sorted_fse.security_lvl != SecurityLvl::Dirty || current_step.is_empty() {
-                current_step.push(sorted_fse);
-            } else if sorted_fse.security_lvl == SecurityLvl::Dirty {
-                // R#1: there only can be max 1 dirty side effect at each step.
-                full_side_effects_steps.push(vec![sorted_fse])
+        > = vec![vec![], vec![]];
+
+        // Split for 2 following steps of Escrow and Optimistic and
+        //  assign indices to each SFX of current Xtx to avoid collisions the same SFX within the same Xtx.
+        for mut sorted_fsx in full_side_effects.iter_mut() {
+            if sorted_fsx.security_lvl == SecurityLvl::Escrow {
+                // ToDo: Replace nonce as field with nonce as Xtx index argument.
+                sorted_fsx.input.nonce = full_side_effects_steps[0].len() as u32;
+                full_side_effects_steps[0].push(sorted_fsx.clone());
+            } else if sorted_fsx.security_lvl == SecurityLvl::Optimistic {
+                sorted_fsx.input.nonce = full_side_effects_steps[1].len() as u32;
+                full_side_effects_steps[1].push(sorted_fsx.clone());
             }
         }
 
+        // full_side_effects_steps should be non-empty at this point
+        if full_side_effects_steps[0].is_empty() {
+            full_side_effects_steps = vec![full_side_effects_steps[1].clone()];
+        } else if full_side_effects_steps[1].is_empty() {
+            full_side_effects_steps = vec![full_side_effects_steps[0].clone()];
+        }
```

Comment:
We should use the safe index function here, `.get(index)`

Reply:
FSX_steps array is declared with 2 elements so don't see why to access assuming could be Option with get

Reply:
Mostly because if this is a bug it doesn't fail gracefully and could endlessly panic the executor.

In this case, you can easily safely destruct the side effect steps with something like `if let (Some(n), Some(n)) = (fss.get(0), fss.get(1) { ... }` and you would be able to log better

Reply:
ok will do

Reply:
Solved by introducing 2 separate vectors  for optimistic and  escrow SFX steps #6da5a9b